core/src/main/java/io/questdb/cutlass/text/TextImportTask.java

```diff
@@ -0,0 +1,995 @@
+                    long lengthAndOffset = Unsafe.getUnsafe().getLong(address + i * 2L * Long.BYTES + Long.BYTES);
+                    long lineLength = lengthAndOffset >>> 48;
+                    offset = lengthAndOffset & MASK;
+                    bytesToRead = lineLength;
+                    additionalLines = 0;
+
+                    for (long j = i + 1; j < count; j++) {
+                        long nextLengthAndOffset = Unsafe.getUnsafe().getLong(address + j * 2L * Long.BYTES + Long.BYTES);
+                        long nextLineLength = nextLengthAndOffset >>> 48;
+                        long nextOffset = nextLengthAndOffset & MASK;
+
+                        //line indexing stops on first EOL char, e.g. \r, but it could be followed by \n
+                        long diff = nextOffset - offset - bytesToRead;
+                        if (diff > -1 && diff < 2 && bytesToRead + nextLineLength < bufLength) {
+                            bytesToRead += diff + nextLineLength;
+                            additionalLines++;
+                        } else {
+                            break;
+                        }
+                    }
+
+                    i += additionalLines;
+
+                    long n = ff.read(fd, buf, bytesToRead, offset);
+                    if (n > 0) {
+                        lexer.parse(buf, buf + n, Integer.MAX_VALUE, listener);
+                    } else {
+                        throw TextException.$("Can't read from file path='").put(path).put("', errno=").put(ff.errno()).put(",offset=").put(offset);
+                    }
+                }
+
+                lexer.parseLast();
+            } finally {
+                if (fd > -1) {
+                    ff.close(fd);
+                }
+
+                Unsafe.free(buf, bufLength, MemoryTag.NATIVE_DEFAULT);
+            }
+        }
+
+        private void mergePartitionIndexAndImportData(final FilesFacade ff,
+                                                      final Path partitionPath,
+                                                      final TextLexer lexer) throws TextException {
+            try (DirectLongList mergeIndexes = new DirectLongList(64, MemoryTag.NATIVE_DEFAULT)) {
+                partitionPath.slash$();
+                int partitionLen = partitionPath.length();
+
+                long mergedIndexSize = openIndexChunks(ff, partitionPath, mergeIndexes, partitionLen);
+                long address = -1;
+                long fd = -1;
+
+                try {
+                    final int indexesCount = (int) mergeIndexes.size() / 2;
+                    partitionPath.trimTo(partitionLen);
+                    partitionPath.concat(CsvFileIndexer.INDEX_FILE_NAME).$();
+
+                    fd = TableUtils.openFileRWOrFail(ff, partitionPath, CairoConfiguration.O_NONE);
+                    address = TableUtils.mapRW(ff, fd, mergedIndexSize, MemoryTag.MMAP_PARALLEL_IMPORT);
+                    ff.close(fd);
```

Comment:
looks like=k spaghetti, yes, the mmaped address can be used when file is closed, but the array of close() methods look very confusing. We can simplify this.

Reply:
Improved